In [51]:
from sklearn import tree
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [52]:
df = pd.read_csv("../Resources/cleaned_data.csv")
df

koi_disposition  koi_fpflag_nt  koi_fpflag_ss  koi_fpflag_co  \
0          CONFIRMED              0              0              0   
1     FALSE POSITIVE              0              1              0   
2     FALSE POSITIVE              0              1              0   
3          CONFIRMED              0              0              0   
4          CONFIRMED              0              0              0   
...              ...            ...            ...            ...   
6986  FALSE POSITIVE              0              0              0   
6987  FALSE POSITIVE              0              1              1   
6988       CANDIDATE              0              0              0   
6989  FALSE POSITIVE              0              0              1   
6990  FALSE POSITIVE              0              0              1   

      koi_fpflag_ec  koi_period  koi_period_err1  koi_period_err2  \
0                 0   54.418383     2.479000e-04    -2.479000e-04   
1                 0   19.899140     1.490000e-05    -1.490000e-05   
2                 0    1.736952     2.630000e-07    -2.630000e-07   
3                 0    2.525592     3.760000e-06    -3.760000e-06   
4                 0    4.134435     1.050000e-05    -1.050000e-05   
...             ...         ...              ...              ...   
6986              1    8.589871     1.846000e-04    -1.846000e-04   
6987              0    0.527699     1.160000e-07    -1.160000e-07   
6988              0    1.739849     1.780000e-05    -1.780000e-05   
6989              0    0.681402     2.430000e-06    -2.430000e-06   
6990              1    4.856035     6.360000e-05    -6.360000e-05   

      koi_time0bk  koi_time0bk_err1  ...  koi_steff_err2  koi_slogg  \
0      162.513840          0.003520  ...             -81      4.467   
1      175.850252          0.000581  ...            -176      4.544   
2      170.307565          0.000115  ...            -174      4.564   
3      171.595550          0.001130  ...            -211      4.438   
4      172.979370          0.001900  ...            -232      4.486   
...           ...               ...  ...             ...        ...   
6986   132.016100          0.015700  ...            -152      4.296   
6987   131.705093          0.000170  ...            -166      4.529   
6988   133.001270          0.007690  ...            -220      4.444   
6989   132.181750          0.002850  ...            -236      4.447   
6990   135.993300          0.010800  ...            -225      4.385   

      koi_slogg_err1  koi_slogg_err2  koi_srad  koi_srad_err1  koi_srad_err2  \
0              0.064          -0.096     0.927          0.105         -0.061   
1              0.044          -0.176     0.868          0.233         -0.078   
2              0.053          -0.168     0.791          0.201         -0.067   
3              0.070          -0.210     1.046          0.334         -0.133   
4              0.054          -0.229     0.972          0.315         -0.105   
...              ...             ...       ...            ...            ...   
6986           0.231          -0.189     1.088          0.313         -0.228   
6987           0.035          -0.196     0.903          0.237         -0.079   
6988           0.056          -0.224     1.031          0.341         -0.114   
6989           0.056          -0.224     1.041          0.341         -0.114   
6990           0.054          -0.216     1.193          0.410         -0.137   

             ra        dec  koi_kepmag  
0     291.93423  48.141651      15.347  
1     297.00482  48.134129      15.436  
2     285.53461  48.285210      15.597  
3     288.75488  48.226200      15.509  
4     296.28613  48.224670      15.714  
...         ...        ...         ...  
6986  298.74921  46.973351      14.478  
6987  297.18875  47.093819      14.082  
6988  286.50937  47.163219      14.757  
6989  294.16489  47.176281      15.385  
6990  297.00977  47.121021      14.826  

[6991 rows x 41 columns]

In [53]:
y = df['koi_disposition']
target_names = ["CONFIRMED", "FALSE POSITIVE", "CANDIDATE"]
X = df.drop(columns='koi_disposition')
feature_names = X.columns
X.head()

koi_fpflag_nt  koi_fpflag_ss  koi_fpflag_co  koi_fpflag_ec  koi_period  \
0              0              0              0              0   54.418383   
1              0              1              0              0   19.899140   
2              0              1              0              0    1.736952   
3              0              0              0              0    2.525592   
4              0              0              0              0    4.134435   

   koi_period_err1  koi_period_err2  koi_time0bk  koi_time0bk_err1  \
0     2.479000e-04    -2.479000e-04   162.513840          0.003520   
1     1.490000e-05    -1.490000e-05   175.850252          0.000581   
2     2.630000e-07    -2.630000e-07   170.307565          0.000115   
3     3.760000e-06    -3.760000e-06   171.595550          0.001130   
4     1.050000e-05    -1.050000e-05   172.979370          0.001900   

   koi_time0bk_err2  ...  koi_steff_err2  koi_slogg  koi_slogg_err1  \
0         -0.003520  ...             -81      4.467           0.064   
1         -0.000581  ...            -176      4.544           0.044   
2         -0.000115  ...            -174      4.564           0.053   
3         -0.001130  ...            -211      4.438           0.070   
4         -0.001900  ...            -232      4.486           0.054   

   koi_slogg_err2  koi_srad  koi_srad_err1  koi_srad_err2         ra  \
0          -0.096     0.927          0.105         -0.061  291.93423   
1          -0.176     0.868          0.233         -0.078  297.00482   
2          -0.168     0.791          0.201         -0.067  285.53461   
3          -0.210     1.046          0.334         -0.133  288.75488   
4          -0.229     0.972          0.315         -0.105  296.28613   

         dec  koi_kepmag  
0  48.141651      15.347  
1  48.134129      15.436  
2  48.285210      15.597  
3  48.226200      15.509  
4  48.224670      15.714  

[5 rows x 40 columns]

The performance of decision tree model

In [54]:
clf = tree.DecisionTreeClassifier()
clf = clf.fit(X_train, y_train)
clf.score(X_test, y_test)

0.8592677345537757

# Pre-processing

Training/test set splitting and one-hot-encoding 

In [55]:
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import GridSearchCV, cross_val_score
from sklearn import metrics
from sklearn.preprocessing import LabelEncoder
from keras.utils import to_categorical

X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1, stratify = y)
label_encoder = LabelEncoder()
label_encoder.fit(y_train)
encoded_y_train = label_encoder.transform(y_train)
encoded_y_test = label_encoder.transform(y_test)
y_train_categorical = to_categorical(encoded_y_train)
y_test_categorical = to_categorical(encoded_y_test)

The performance of random forest model with default setting on training set

In [56]:
rf0 = RandomForestClassifier(oob_score=True, random_state=10)
rf0.fit(X_train,encoded_y_train)
print(rf0.oob_score_)
y_predprob = rf0.predict_proba(X_train) # the probabilities of each class
print(f"AUC Score (Train):{metrics.roc_auc_score(encoded_y_train, rf0.predict_proba(X_train), multi_class='ovr')}")

0.8928094602326913
AUC Score (Train):1.0


# Model Tuning
First, optimize the model by tuning the number of estimators (number of trees in forest)

In [57]:
param_test1 = {'n_estimators':range(10,201,10)}
gsearch1 = GridSearchCV(estimator = RandomForestClassifier(min_samples_split=100,
                                  min_samples_leaf=20, max_depth=8,max_features='sqrt',random_state=10), 
                       param_grid = param_test1, scoring='roc_auc',cv=5)

gsearch1.fit(X_train, y_train_categorical)
gsearch1.cv_results_['mean_test_score'], gsearch1.best_params_, gsearch1.best_score_

(array([0.94834299, 0.95574062, 0.95726427, 0.95810021, 0.95861513,
        0.95852297, 0.95884636, 0.95902091, 0.95889067, 0.95873874,
        0.95909634, 0.95946837, 0.9593618 , 0.95947955, 0.95960087,
        0.95943025, 0.95948002, 0.95974088, 0.95987783, 0.95981441]),
 {'n_estimators': 190},
 0.959877825281691)

According to the optimization results, the best hyperparameter for estimator is 190. The resulted cross-validated AUC is 95.99%.

Then, optimze the model by tuning the max_depth

In [58]:
param_test2 = {'max_depth':range(3,14,2), 'min_samples_split':range(50,201,20)}
gsearch2 = GridSearchCV(estimator = RandomForestClassifier(n_estimators = 190,
                                  min_samples_leaf=20,max_features='sqrt',random_state=10), 
                       param_grid = param_test2, scoring='roc_auc',iid=False,cv=5)

gsearch2.fit(X_train, y_train_categorical)
gsearch2.cv_results_['mean_test_score'], gsearch2.best_params_, gsearch2.best_score_

KeyboardInterrupt: 

In [47]:
rf1 = RandomForestClassifier(n_estimators= 60, max_depth=13, min_samples_split=110,
                                  min_samples_leaf=20,max_features='sqrt' ,oob_score=True, random_state=10)
rf1.fit(X,y)
print rf1.oob_score_

In [47]:
param_test3 = {'min_samples_split':range(80,150,20), 'min_samples_leaf':range(10,60,10)}
gsearch3 = GridSearchCV(estimator = RandomForestClassifier(n_estimators= 60, max_depth=13,
                                  max_features='sqrt' ,oob_score=True, random_state=10),
   param_grid = param_test3, scoring='roc_auc',iid=False, cv=5)
gsearch3.fit(X_train, y_train_categorical)
gsearch3.grid_scores_, gsearch3.best_params_, gsearch3.best_score_

In [47]:
param_test4 = {'max_features':range(3,11,2)}
gsearch4 = GridSearchCV(estimator = RandomForestClassifier(n_estimators= 60, max_depth=13, min_samples_split=120,
                                  min_samples_leaf=20 ,oob_score=True, random_state=10),
   param_grid = param_test4, scoring='roc_auc',iid=False, cv=5)

gsearch4.fit(X_train, y_train_categorical)
gsearch4.grid_scores_, gsearch4.best_params_, gsearch4.best_score_

In [47]:
rf2 = RandomForestClassifier(n_estimators= 60, max_depth=13, min_samples_split=120,
                                  min_samples_leaf=20,max_features=7 ,oob_score=True, random_state=10)
rf2.fit(X,y)
print rf2.oob_score_

In [47]:
sorted(zip(rf.feature_importances_, feature_names), reverse=True)

[(0.11133271975468742, 'koi_fpflag_co'),
 (0.09449208195590787, 'koi_fpflag_nt'),
 (0.07333934602403787, 'koi_fpflag_ss'),
 (0.056327167343656614, 'koi_model_snr'),
 (0.05099226881431779, 'koi_prad'),
 (0.03418475299321093, 'koi_prad_err2'),
 (0.033829653797619916, 'koi_fpflag_ec'),
 (0.033551325543753376, 'koi_duration_err1'),
 (0.03232723774740154, 'koi_duration_err2'),
 (0.030253088269462332, 'koi_steff_err2'),
 (0.028929890967020776, 'koi_prad_err1'),
 (0.02470861513226234, 'koi_steff_err1'),
 (0.023288986850252064, 'koi_duration'),
 (0.02259092281263069, 'koi_period'),
 (0.02124480159914948, 'koi_time0bk_err2'),
 (0.021225712906921528, 'koi_insol_err1'),
 (0.02099995883607792, 'koi_time0bk_err1'),
 (0.01969685175932427, 'koi_depth'),
 (0.018381226311847514, 'koi_period_err2'),
 (0.017414292655411383, 'koi_period_err1'),
 (0.017080786950508572, 'koi_impact'),
 (0.016096248777969238, 'koi_insol_err2'),
 (0.014358692951865369, 'koi_depth_err2'),
 (0.014321729728557175, 'koi_teq'),
 (

In [48]:
import joblib
filename = "random_forest.sav"
joblib.dump(rf, filename)

['random_forest.sav']

In [51]:
predictions = rf.predict(X_test)
predictions_df = pd.DataFrame({"Actual": y_test, "Predicted":predictions})
predictions_df.head(20)

Actual       Predicted
1213       CONFIRMED       CONFIRMED
5220  FALSE POSITIVE  FALSE POSITIVE
6688  FALSE POSITIVE  FALSE POSITIVE
2031       CONFIRMED       CONFIRMED
5824  FALSE POSITIVE  FALSE POSITIVE
2053  FALSE POSITIVE  FALSE POSITIVE
2567  FALSE POSITIVE  FALSE POSITIVE
1687       CONFIRMED       CONFIRMED
1837       CANDIDATE       CANDIDATE
4016  FALSE POSITIVE  FALSE POSITIVE
901        CONFIRMED       CONFIRMED
691   FALSE POSITIVE  FALSE POSITIVE
724   FALSE POSITIVE  FALSE POSITIVE
5011       CANDIDATE       CANDIDATE
2912       CANDIDATE       CANDIDATE
2641       CONFIRMED       CONFIRMED
6588  FALSE POSITIVE  FALSE POSITIVE
5705  FALSE POSITIVE  FALSE POSITIVE
1657       CONFIRMED       CONFIRMED
3004       CANDIDATE       CANDIDATE

In [55]:
predictions_df['match'] = np.where(predictions_df['Predicted'] == predictions_df['Actual'], 1, 0)
print(f"The number of total predictions: {len(predictions_df)}")
print(f"The number of correct predictions: {sum(predictions_df['match'])}")
print(f"The test test accuracy: {round(sum(predictions_df['match'])/len(predictions_df)*100,2)}%")
predictions_df[predictions_df['match']==0].index

The number of total predictions: 1748
The number of correct predictions: 1599
The test test accuracy: 91.48%


Int64Index([5248, 2080,  161,  906, 2246,  119, 1862, 2624, 1316, 2019,
            ...
            2326, 2270,  795, 1288, 1685, 1228,  289, 2271, 2539,  958],
           dtype='int64', length=149)